vibora/pdf2audio.py

```diff
@@ -3,9 +3,18 @@
 
 # pdf to audio
 def speak_text(text):
-  engine = pyttsx3.init()
-  engine.say(text)
-  engine.runAndWait()
+  try:
+    engine = pyttsx3.init()
+  except Exception as e:
+    print(f'{9}: {e}')
+  try:
+    engine.say(text)
+  except Exception as e:
+    print(f'{13}: {e}')
+  try:
+    engine.runAndWait()
+  except Exception as e:
+    print(f'{18}: {e}')
 
```

Comment:
output looked like:

```
$ python main.py pdf2audio Academic\ Transcript.pdf
9: [WinError -2147221008] CoInitialize has not been called
9: [WinError -2147221008] CoInitialize has not been called
13: cannot access local variable 'engine' where it is not associated with a value
9: [WinError -2147221008] CoInitialize has not been called
9: [WinError -2147221008] CoInitialize has not been called
9: [WinError -2147221008] CoInitialize has not been called
9: [WinError -2147221008] CoInitialize has not been called
13: cannot access local variable 'engine' where it is not associated with a value
13: cannot access local variable 'engine' where it is not associated with a value
18: cannot access local variable 'engine' where it is not associated with a value
13: cannot access local variable 'engine' where it is not associated with a value
18: cannot access local variable 'engine' where it is not associated with a value
18: cannot access local variable 'engine' where it is not associated with a value
9: [WinError -2147221008] CoInitialize has not been called
13: cannot access local variable 'engine' where it is not associated with a value
18: cannot access local variable 'engine' where it is not associated with a value
13: cannot access local variable 'engine' where it is not associated with a value
13: cannot access local variable 'engine' where it is not associated with a value
18: cannot access local variable 'engine' where it is not associated with a value
18: cannot access local variable 'engine' where it is not associated with a value
18: cannot access local variable 'engine' where it is not associated with a value
```

Reply:
> output looked like:
> 
> ```
> $ python main.py pdf2audio Academic\ Transcript.pdf
> 9: [WinError -2147221008] CoInitialize has not been called
> 9: [WinError -2147221008] CoInitialize has not been called
> 13: cannot access local variable 'engine' where it is not associated with a value
> 9: [WinError -2147221008] CoInitialize has not been called
> 9: [WinError -2147221008] CoInitialize has not been called
> 9: [WinError -2147221008] CoInitialize has not been called
> 9: [WinError -2147221008] CoInitialize has not been called
> 13: cannot access local variable 'engine' where it is not associated with a value
> 13: cannot access local variable 'engine' where it is not associated with a value
> 18: cannot access local variable 'engine' where it is not associated with a value
> 13: cannot access local variable 'engine' where it is not associated with a value
> 18: cannot access local variable 'engine' where it is not associated with a value
> 18: cannot access local variable 'engine' where it is not associated with a value
> 9: [WinError -2147221008] CoInitialize has not been called
> 13: cannot access local variable 'engine' where it is not associated with a value
> 18: cannot access local variable 'engine' where it is not associated with a value
> 13: cannot access local variable 'engine' where it is not associated with a value
> 13: cannot access local variable 'engine' where it is not associated with a value
> 18: cannot access local variable 'engine' where it is not associated with a value
> 18: cannot access local variable 'engine' where it is not associated with a value
> 18: cannot access local variable 'engine' where it is not associated with a value
> ```

Even with this output was it able to read the file with no problems?
And the thing with `pythoncom` is that is depends on `pywin32`, which is Windows-only.

Reply:
Well that's very weird. Looks like something is wrong with `espeak` but, outside of vibora it is working just fine. Am also getting that `run loop already started` in `pyttsx3`.

Here is the full output:

```
Exception in thread Thread-14 (speak_text):
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
Exception ignored on calling ctypes callback function: <bound method EspeakDriver._onSynth of <pyttsx3.drivers.espeak.EspeakDriver object at 0x7fe3dc997430>>
Traceback (most recent call last):
  File "/home/smooth/fun/vibora/env/lib/python3.10/site-packages/pyttsx3/drivers/espeak.py", line 153, in _onSynth
    self._proxy.notify('started-word',
    self.run()
AttributeError: 'EspeakDriver' object has no attribute '_proxy'
  File "/usr/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/home/smooth/fun/vibora/vibora/pdf2audio.py", line 8, in speak_text
Exception ignored on calling ctypes callback function: <bound method EspeakDriver._onSynth of <pyttsx3.drivers.espeak.EspeakDriver object at 0x7fe3dc997430>>
Traceback (most recent call last):
  File "/home/smooth/fun/vibora/env/lib/python3.10/site-packages/pyttsx3/drivers/espeak.py", line 153, in _onSynth
    engine.runAndWait()
  File "/home/smooth/fun/vibora/env/lib/python3.10/site-packages/pyttsx3/engine.py", line 177, in runAndWait
    self._proxy.notify('started-word',
AttributeError: 'EspeakDriver' object has no attribute '_proxy'
    raise RuntimeError('run loop already started')
RuntimeError: run loop already started
Exception ignored on calling ctypes callback function: <bound method EspeakDriver._onSynth of <pyttsx3.drivers.espeak.EspeakDriver object at 0x7fe3dc997430>>
Exception in thread Thread-15 (speak_text):
Traceback (most recent call last):
Traceback (most recent call last):
  File "/home/smooth/fun/vibora/env/lib/python3.10/site-packages/pyttsx3/drivers/espeak.py", line 153, in _onSynth
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self._proxy.notify('started-word',
AttributeError: 'EspeakDriver' object has no attribute '_proxy'
    self.run()
Exception ignored on calling ctypes callback function: <bound method EspeakDriver._onSynth of <pyttsx3.drivers.espeak.EspeakDriver object at 0x7fe3dc997430>>
  File "/usr/lib/python3.10/threading.py", line 953, in run
Traceback (most recent call last):
Exception in thread Thread-16 (speak_text):
    self._target(*self._args, **self._kwargs)
  File "/home/smooth/fun/vibora/env/lib/python3.10/site-packages/pyttsx3/drivers/espeak.py", line 153, in _onSynth
Exception in thread Thread-17 (speak_text):
Traceback (most recent call last):
Traceback (most recent call last):
  File "/home/smooth/fun/vibora/vibora/pdf2audio.py", line 8, in speak_text
    self._proxy.notify('started-word',
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
Exception in thread Thread-18 (speak_text):
AttributeError: 'EspeakDriver' object has no attribute '_proxy'
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    engine.runAndWait()
    self.run()
Exception in thread Thread-19 (speak_text):
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 953, in run
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
  File "/home/smooth/fun/vibora/env/lib/python3.10/site-packages/pyttsx3/engine.py", line 177, in runAndWait
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
    self._target(*self._args, **self._kwargs)
    raise RuntimeError('run loop already started')
    self.run()
Exception in thread Thread-21 (speak_text):
Traceback (most recent call last):
RuntimeError: run loop already started
  File "/usr/lib/python3.10/threading.py", line 953, in run
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
  File "/usr/lib/python3.10/threading.py", line 953, in run
Exception in thread Thread-22 (speak_text):
Traceback (most recent call last):
    self.run()
    self._target(*self._args, **self._kwargs)
  File "/home/smooth/fun/vibora/vibora/pdf2audio.py", line 8, in speak_text
    self.run()
  File "/usr/lib/python3.10/threading.py", line 953, in run
Exception in thread Thread-23 (speak_text):
Exception in thread Thread-20 (speak_text):
    engine.runAndWait()
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 953, in run
  File "/home/smooth/fun/vibora/env/lib/python3.10/site-packages/pyttsx3/engine.py", line 177, in runAndWait
    self._target(*self._args, **self._kwargs)
    self._target(*self._args, **self._kwargs)
  File "/home/smooth/fun/vibora/vibora/pdf2audio.py", line 8, in speak_text
Exception in thread Thread-25 (speak_text):
    self._target(*self._args, **self._kwargs)
  File "/home/smooth/fun/vibora/vibora/pdf2audio.py", line 8, in speak_text
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
  File "/home/smooth/fun/vibora/vibora/pdf2audio.py", line 8, in speak_text
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.10/threading.py", line 953, in run
    engine.runAndWait()
Exception in thread Thread-24 (speak_text):
Traceback (most recent call last):
Exception in thread Thread-27 (speak_text):
    engine.runAndWait()
  File "/home/smooth/fun/vibora/vibora/pdf2audio.py", line 8, in speak_text
Exception in thread Thread-26 (speak_text):
Traceback (most recent call last):
  File "/home/smooth/fun/vibora/env/lib/python3.10/site-packages/pyttsx3/engine.py", line 177, in runAndWait

    raise RuntimeError('run loop already started')
    self.run()
  File "/usr/lib/python3.10/threading.py", line 953, in run
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self._target(*self._args, **self._kwargs)
Traceback (most recent call last):
  File "/home/smooth/fun/vibora/env/lib/python3.10/site-packages/pyttsx3/engine.py", line 177, in runAndWait
    engine.runAndWait()
  File "/home/smooth/fun/vibora/env/lib/python3.10/site-packages/pyttsx3/engine.py", line 177, in runAndWait
    self._target(*self._args, **self._kwargs)
  File "/home/smooth/fun/vibora/vibora/pdf2audio.py", line 8, in speak_text
Fatal Python error: _enter_buffered_busy: could not acquire lock for <_io.BufferedWriter name='<stderr>'> at interpreter shutdown, possibly due to daemon threads
Python runtime state: finalizing (tstate=0x000055e756ca7a90)

Thread 0x00007fe3baffd640 (most recent call first):
  <no Python frame>

Current thread 0x00007fe3e14671c0 (most recent call first):
  <no Python frame>

Extension modules: psutil._psutil_linux, psutil._psutil_posix, PIL._imaging, fitz._fitz, PIL._imagingcms, lxml._elementpath, lxml.etree, getch (total: 8)
```

Reply:
got it, i was able to dual boot ubuntu so i'll try to replicate the error tonight

Reply:
which distro and release are you running? I haven't been able to get pdf2audio running at all due to dependencies and python versioning, so I'll try a VM next with your build. I'm sorry for my back-and-forth!

Reply:
My bad, I should have told you that hahaha, anyway...

I'm running Linux Mint 21.1, I don't know if the DE could have an impact based on the packages installed but, I'm using Cinnamon. And running kernel 5.15.0-75-generic.

This is a fresh install, I formatted it after I started getting these errors, to see if it would not happen with a fresh build. So you'll pretty much have the same system build running as me.

Reply:
I was able to get it working by removing `threading`. It looks like both `threading` and `keyboard` require root permission on Linux, so I removed `threading` and added `getch` instead of `keyboard`.

A few things to mention:
- The voice (at least here on Linux) is very robotic. Maybe it's because of `espeak`?
- I needed to press "space" after it read the first page of `testfiles/testpaper1.pdf` to start reading the second page, code didn't break, just had to press it in order to start reading the next page.

If it's possible, let me know if you can try it on Windows to see if everything is working fine with these changes.
Also, test it using the testpaper1.pdf in testfiles/ just like I did.

Here is the modified code:

```python
import pyttsx3
import time
from PyPDF2 import PdfReader
from getch import getch

# pdf to audio
def speak_text(text):
    engine = pyttsx3.init()
    engine.say(text)
    engine.runAndWait()

def audio(pdf_path):
    with open(pdf_path, 'rb') as f:
        pdf_reader = PdfReader(f)
        for page in pdf_reader.pages:
            text = page.extract_text()
            speak_text(text)
            sleep_duration = 0.5
            start_time = time.time()

            while time.time() - start_time < sleep_duration:
                if getch() == b'\x03':  # Check for Ctrl+C key press
                    return  # Exit the function if Ctrl+C is pressed
                else:
                    continue # to go to the next page automatically
```

Reply:
i'd have to change it a bit to run, using `msvcrt.getch()`

https://pypi.org/project/getch/

Reply:
Could you test this code to see if it'll work?

I used `platform` to detect which OS the user is using

```python
import pyttsx3
import time
from PyPDF2 import PdfReader
import platform

if platform.system() == "Windows":
    from msvcrt import getch
else:
    from getch import getch

# pdf to audio
def speak_text(text):
    engine = pyttsx3.init()
    engine.say(text)
    engine.runAndWait()

def audio(pdf_path):
    with open(pdf_path, 'rb') as f:
        pdf_reader = PdfReader(f)
        for page in pdf_reader.pages:
            text = page.extract_text()
            speak_text(text)
            sleep_duration = 0.5
            start_time = time.time()

            while time.time() - start_time < sleep_duration:
                if getch() == b'\x03':  # Check for Ctrl+C key press
                    return  # Exit the function if Ctrl+C is pressed
                else:
                    continue # to go to the next page automatically
```

Reply:
yes, this works as you described:
- it pauses between pages
- if i press Ctrl+C it exits gracefully
- if i press anything else, it continues reading

Reply:
Would you mind comitting these changes we made? I'll merge it later.
I'll also take a look at these pauses between pages in another time, but at least we have it working for now.